boot/bootstate20.go

```diff
@@ -258,93 +206,386 @@ func (u20 *bootStateUpdate20) commit() error {
+		// kernel, and got rebooted before the kernel_status was updated, we would
+		// have kernel_status="trying" which would cause the bootloader to think
+		// the boot failed, and revert to booting using the kernel symlink, but that
+		// now points to the new kernel we were trying and we did not successfully
+		// boot from that kernel to know we should trust it.
+		// The try-kernel symlink removal should happen last because it will not
+		// affect anything, except that if it was removed before updating
+		// kernel_status to "", the bootloader will think that the try kernel failed
+		// to boot and fall back to booting the old kernel which is safe.
+
+		// clear the kernel_status boot var first
+		err := bsmark.ebl.SetBootVars(map[string]string{"kernel_status": ""})
 		if err != nil {
 			return err
 		}
-	}
 
-	// remove try kernel symlink
-	// * step 3 for successful boot after an update
-	// * NOP for about to try an update
-	if u20.env["kernel_status"] != "" {
-		err = u20.ebl.DisableTryKernel()
+		// enable the kernel we tried
+		err = bsmark.ebl.EnableKernel(bsmark.triedKernelSnap)
 		if err != nil {
 			return err
 		}
+
+		// finally disable the try kernel symlink
+		err = bsmark.ebl.DisableTryKernel()
+		if err != nil {
+			return err
+		}
+	} else {
+		// triedKernelSnap is not true, but we still might need to clean up
+		// kernel_status, so check if what we are to commit is different from
+		// what we read
+		if bsmark.commitKernelStatus != bsmark.kernelStatus {
+			m := map[string]string{
+				"kernel_status": bsmark.commitKernelStatus,
+			}
+
+			// set the boot variables
+			err := bsmark.ebl.SetBootVars(m)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// base snap next
+	if bsmark.triedBaseSnap != nil {
+		// the ordering here is less important, since the only operation that
+		// has side-effects is writing the modeenv at the end, and that uses an
+		// atomic file writing operation, so it's not a concern if we get
+		// rebooted during this snippet like it is with the kernel snap above
+
+		changed := false
+
+		// clear the status
+		if bsmark.modeenv.BaseStatus != "" {
+			changed = true
+			bsmark.modeenv.BaseStatus = ""
+		}
+		// set the new base as the tried base snap
+		tryBase := filepath.Base(bsmark.triedBaseSnap.MountFile())
+		if bsmark.modeenv.Base != tryBase {
+			bsmark.modeenv.Base = tryBase
+			changed = true
+		}
+
+		// clear the TryBase
+		if bsmark.modeenv.TryBase != "" {
+			bsmark.modeenv.TryBase = ""
+			changed = true
+		}
+
+		// write the modeenv
+		if changed {
+			return bsmark.modeenv.Write("")
+		}
 	}
 
 	return nil
 }
+
+// genericSetNext implements the generic logic for setting up a snap to be tried
+// for boot and works for both kernel and base snaps (though not
+// simultaneously).
+func genericSetNext(b bootState, currentStatus string, next snap.PlaceInfo) (rebootRequired bool, setAsTrySnap bool, setStatus string, err error) {
```

Comment:
I think we can simplify things a bit by changing the return signature a bit more:

afaict   rebootRequired is true iff seAsTrySnap is true?


we can return  trySnap snap.PlaceInfo that can be nil or next instead of the two flags

Reply:
okay, I went a bit further and now only return status and and error here, then the specific setNext() methods check if status is "try", and set the snap to be tried and return rebootRequired=true.